### DISCUSSION TITLE
pass dictionary key to member model

### DISCUSSION THREAD

**Discussion by phemmer** [pydantic/pydantic#3003] [Question]
Lets say I have a model where I want a nested model to be indexed within a dictionary by some attribute of that child model. For example:
```python
class Book(BaseModel):
  title:str
  published_year:int

class Author(BaseModel):
  name:str
  books:Dict[str,Book]
```

This would require repetition of the book's title when constructing the data. For example a serialized `Author` object might look like:
```json
{
  "name": "Herman Melville",
  "books": {
    "Moby-Dick": {
      "title": "Moby-Dick",
      "published_year": 1851
    }
  }
}
```

Instead of having to repeat the book's title, is there any way to omit the explicit `title` field and use the dictionary key from `books` as the value?

----

I've tried creating a solution using `@validator('books', pre=True)`, and having it iterate over the `books` dictionary, taking each key and injecting it into the nested dictionary as the value of key `title`. This works, but when generating the JSON schema, it lists `title` as a required field for `Book`. I can't make the attribute private, as then injecting the value in the validator wouldn't work because pydantic won't let you set a private attribute during construction.

**Comment by PrettyWood** [comment]
Hi @phemmer 
I'm not sure I really understood the exact behaviour you want.
Sorry if I misunderstood
```py
from pydantic import BaseModel, validator


class Book(BaseModel):
    title: str
    published_year: int

    def dict(self, *, exclude = None, **kwargs) -> dict[str, object]:
        if exclude is None:
            exclude = {"title"}
        return super().dict(exclude=exclude, **kwargs)


class Author(BaseModel):
    name:str
    books: dict[str, Book]

    @validator("books", pre=True)
    def add_title(cls, value: dict[str, dict[str, object]]) -> dict[str, dict[str, object]]:
        for title, book_infos in value.items():
            book_infos["title"] = title
        return value


payload = {
  "name": "Herman Melville",
  "books": {
    "Moby-Dick": {
      "published_year": 1851
    },
    "Pika": {
      "published_year": 3000
    }
  }
}
author = Author(**payload)
assert repr(author) == "Author(name='Herman Melville', books={'Moby-Dick': Book(title='Moby-Dick', published_year=1851), 'Pika': Book(title='Pika', published_year=3000)})"
assert author.dict() == {"name": "Herman Melville", "books": {"Moby-Dick": {"published_year": 1851}, "Pika": {"published_year": 3000}}}
```
Hope it helps
